ENH: Providing ITKv5 updates for C++11

hjmjohnson opened the pull request:
This pull request incorporates code changes
that happened in parallel with updating the
core ITK code to ITKv5 for C++11 conformity.

hjmjohnson committed:
STYLE: Replace itkStaticConstMacro with static constexpr

Use static constexpr directly now that C++11 conformance
is required by all compilers.

:%s/itkStaticConstMacro *( *\([^,]*\),[ \_s]*\([^,]*\),\_s*\([^)]*\)) */static constexpr \2 \1 = \3/ge

hjmjohnson committed:
STYLE: Prefer C++11 type alias over typedef

== http://en.cppreference.com/w/cpp/language/type_alias ==

Type alias is a name that refers to a previously defined type (similar
to typedef).

A type alias declaration introduces a name which can be used as a
synonym for the type denoted by type-id. It does not introduce a new
type and it cannot change the meaning of an existing type name. There is
no difference between a type alias declaration and typedef declaration.
This declaration may appear in block scope, class scope, or namespace
scope.

== https://www.quora.com/Is-using-typedef-in-C++-considered-a-bad-practice ==

While typedef is still available for backward compatibility, the new
Type Alias syntax 'using Alias = ExistingLongName;' is more consistent
with the flow of C++ than the old typedef syntax 'typedef
ExistingLongName Alias;', and it also works for templates (Type alias,
alias template (since C++11)), so leftover 'typedef' aliases will differ
in style from any alias templates.

dzenanz commented on include/itkCartesianToPolarTransform.h:
```diff
@@ -1,184 +1,184 @@
+  itkTypeMacro( CartesianToPolarTransform, Transform );
+
+  /** Dimension of the domain space. */
+  static constexpr unsigned int SpaceDimension = NDimensions;
```
Do we want `static` here?

hjmjohnson replied:
That what was there before.  Perhaps not, but I'm not trying to maintain this code at the same level as ITK proper.   Feel free to make changes as you see fit.

PS:  Thanks for you reviews.

dzenanz commented on include/itkCartesianToPolarTransform.h:
```diff
@@ -1,184 +1,184 @@
+   * This method transforms first two dimensions of a point from cartesian
+   * coordinates to polar coordinates <alpha,radius>.
+   */
+  OutputPointType TransformPoint(const InputPointType  &point ) const ITK_OVERRIDE;
```
ITK_OVERRIDE -> override ?

hjmjohnson committed:
COMP:  Use C++11 override directly

git grep -l "ITK_OVERRIDE" |   fgrep -v itk_compiler_detection.h | fgrep -v CMakeLists.txt |fgrep -v .cmake |   xargs sed -i '' -e "s/ITK_OVERRIDE/override/g"

hjmjohnson committed:
STYLE: Prefer constexpr for const numeric literals

Use constexpr for constant numeric literals.
